### PR TITLE
Hc 56 add select city for the clinic functionality

### PR DESCRIPTION
**Pull request by ReznikYuriy**

### CODE REVIEW

File: backend/src/data/repositories/city-repository.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+  public getAll(): Promise<ICity[]> {
+    return CityModel.findAll();
+  }
+
+  public createCity(cityName: Partial<ICity>): Promise<ICity> {
+    return CityModel.create(cityName);
+  }
+}
+
+export { CityRepository };
```
Comment (ghost) [inline backend/src/data/repositories/city-repository.ts]:
```suggestion
class City {

  public getAll(): Promise<ICity[]> {
    return CityModel.findAll();
  }

  public createCity(cityName: Partial<ICity>): Promise<ICity> {
    return CityModel.create(cityName);
  }
}

export { City };
```
Rename file to `backend/src/data/repositories/city.repository.ts`
Rename all repositories this way

File: frontend/src/components/clinics/components/add-clinic-popup/add-clinic-popup.tsx [hunk 2]
```diff
@@ -30,16 +33,36 @@ const AddClinicPopup: React.FC<IProps> = ({
...
+
+  const dispatch = useDispatch();
+
+  const [selectInputValue, setSelectInputValue] = React.useState<string>('');
+
+  React.useEffect(() => {
+    dispatch(ClinicsActionCreator.getCities());
+  }, []);
+
+  const cityList: IOption[] = cities.map(city => ({ label: city.name, value: city.id }));
```
Comment (ghost) [inline frontend/src/components/clinics/components/add-clinic-popup/add-clinic-popup.tsx]: use `createOptions` helpers

File: frontend/src/components/clinics/components/common/constants/default-clinic-value.constant.ts [hunk 3]
```diff
@@ -7,6 +7,7 @@ const DEFAULT_CLINIC_VALUE: IClinicPayload = {
   [ClinicKey.ADDRESS]: '',
   [ClinicKey.IMAGE_PATH]: ClinicDefaultImage,
   [ClinicKey.CLINIC_TYPE]: ClinicType.STATE,
+  [ClinicKey.CITY_ID]:' ',
```
Comment (ghost) [inline frontend/src/components/clinics/components/common/constants/default-clinic-value.constant.ts]: Just make it nullable
```suggestion
  [ClinicKey.CITY_ID]: null,
```

File: frontend/src/services/city-api/city-api.service.ts [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+  #http: Http;
+  #apiPrefix: string;
+
+  constructor({ http, apiPrefix }: Constructor) {
+    this.#http = http;
+    this.#apiPrefix = apiPrefix;
+  }
+
+  public addCity(payload: Partial<ICity>): Promise<ICity> {
+    return this.#http.load(`${this.#apiPrefix}${ApiPath.CITIES}${CityApiPath.ROOT}`, {
```
Comment (ghost) [inline frontend/src/services/city-api/city-api.service.ts]:
```suggestion
    return this.#http.load(`${this.#apiPrefix}${ApiPath.CITIES}`, {
```

File: frontend/src/services/city-api/city-api.service.ts [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+      method: HttpMethod.POST,
+      contentType: ContentType.JSON,
+      payload,
+    });
+  }
+
+  public getCities(): Promise<ICity[]> {
+    return this.#http.load(
+      `${this.#apiPrefix}${ApiPath.CITIES}${CityApiPath.ROOT}`,
```
Comment (ghost) [inline frontend/src/services/city-api/city-api.service.ts]:
```suggestion
      `${this.#apiPrefix}${ApiPath.CITIES}`,
```

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/store/slices/clinics/clinics.slice.ts [hunk 6]
```diff
@@ -38,8 +50,16 @@ const getClinics = (): AppThunk => async (dispatch) => {
...
+) => {
   try {
+    if (clinicInfo.cityId === '' && cityValue !== '') {
+      const response = await cityApi.addCity({ name: cityValue });
+      dispatch(actions.addCity(response));
+      clinicInfo.cityId=response.id;
+    }
+
     const response = await clinicApi.addClinic(clinicInfo);
     dispatch(actions.addClinic([response]));
```
Comment (ghost) [inline frontend/src/store/slices/clinics/clinics.slice.ts]:
```suggestion
    dispatch(actions.addClinic(response));
```

File: frontend/src/store/slices/clinics/clinics.slice.ts [hunk 7]
```diff
@@ -23,6 +29,12 @@ const { reducer, actions } = createSlice({
     addClinic: (state, action: PayloadAction<IClinic[]>) => {
       state.clinics = [...state.clinics, ...action.payload];
```
Comment (ghost) [inline frontend/src/store/slices/clinics/clinics.slice.ts]:
```suggestion
    addClinic: (state, action: PayloadAction<IClinic>) => {
      state.clinics = [...state.clinics, action.payload];
```

File: frontend/src/store/slices/clinics/clinics.slice.ts [hunk 8]
```diff
@@ -38,8 +50,16 @@ const getClinics = (): AppThunk => async (dispatch) => {
...
-const addClinic = (clinicInfo: IClinicPayload): AppThunk => async (dispatch) => {
+const addClinic = (clinicInfo: IClinicPayload, cityValue?: string): AppThunk => async (
+  dispatch,
+) => {
   try {
+    if (clinicInfo.cityId === '' && cityValue !== '') {
+      const response = await cityApi.addCity({ name: cityValue });
+      dispatch(actions.addCity(response));
+      clinicInfo.cityId=response.id;
+    }
```
Comment (ghost) [inline frontend/src/store/slices/clinics/clinics.slice.ts]:
```suggestion
  const updatedClinic: IClinic = {
    ... clinicInfo
  }
  try {
    if (clinicInfo.cityId === '' && cityValue !== '') {
      const response = await cityApi.addCity({ name: cityValue });
      dispatch(actions.addCity(response));
      updatedClinic.cityId = response.id;
    }
```

File: shared/src/common/enums/api/api-path.enum.ts [hunk 9]
```diff
@@ -9,6 +9,8 @@ enum ApiPath {
   DIAGNOSES = '/diagnoses',
   FILES = '/files',
   PERMISSIONS = '/permissions',
+  MAP = '/map',
```
Comment (ghost) [inline shared/src/common/enums/api/api-path.enum.ts]:
```suggestion
```

File: shared/src/validation-schemas/clinic/clinic.validation-schema.ts [hunk 10]
```diff
@@ -16,10 +16,8 @@ const addClinic = yup.object().shape({
     .mixed<ClinicType>()
     .oneOf(Object.values(ClinicType))
     .required(),
-//   [ClinicKey.IMAGE_PATH]: yup
-//     .string()
-//     .required(RegisterValidationMessage.IMAGE_REQUIRED)
-//     .url(RegisterValidationMessage.IMAGE_INCORRECT),
+  [ClinicKey.CITY_ID]: yup
+    .string(),
```
Comment (ghost) [inline shared/src/validation-schemas/clinic/clinic.validation-schema.ts:20]:
```suggestion
    .string(),
    .url(RegisterValidationMessage.IMAGE_INCORRECT)
```
Remove it if we don't use it `RegisterValidationMessage.IMAGE_REQUIRED`

**Review by ghost — CHANGES_REQUESTED** [review]

File: shared/src/validation-schemas/clinic/clinic.validation-schema.ts (hunk 10 above)
Comment (ZubkoWhatislove) [reply shared/src/validation-schemas/clinic/clinic.validation-schema.ts:20]: Missed

**Review by ZubkoWhatislove — CHANGES_REQUESTED** [review]

**Pull request merged**